nomad/job_endpoint_hooks.go

```diff
@@ -206,14 +211,58 @@ func (jobImpliedConstraints) Mutate(j *structs.Job) (*structs.Job, []error, erro
 		}
 
 		// If the task group utilises Consul service discovery, run the mutator.
-		if ok := consulServiceDisco[tg.Name]; ok {
-			mutateConstraint(constraintMatcherLeft, tg, consulServiceDiscoveryConstraint)
+		if consulServiceDisco[tg.Name] {
+			for _, service := range tg.Services {
+				if service.IsConsul() {
+					mutateConstraint(constraintMatcherLeft, tg, consulConstraintFn(service))
+				}
+
+				for _, task := range tg.Tasks {
+					for _, service := range task.Services {
+						if service.IsConsul() {
+							mutateConstraint(constraintMatcherLeft, tg, consulConstraintFn(service))
+						}
+					}
+				}
+			}
 		}
+
 	}
 
 	return j, nil, nil
 }
 
+// vaultConstraintFn returns a constraint that matches the fingerprint of the
+// requested Vault cluster. This is to support Nomad Enterprise but neither the
+// fingerprint or non-default cluster are allowed well before we get here, so no
+// need to split out the behavior to ENT-specific code.
+func vaultConstraintFn(vault *structs.Vault) *structs.Constraint {
+	if vault.Cluster != "default" && vault.Cluster != "" {
+		return &structs.Constraint{
+			LTarget: fmt.Sprintf("${attr.vault.%s.version}", vault.Cluster),
+			RTarget: ">= 0.6.1",
+			Operand: structs.ConstraintSemver,
+		}
```

Comment:
Would this generate a confusing placement error message? As a job submitter, I usually don't care about Vault version 😅 

Perhaps something like this generates a more meaningful error?

```suggestion
		return &structs.Constraint{
			LTarget: fmt.Sprintf("${attr.vault.%s}", vault.Cluster),
			Operand: structs.ConstraintAttributeIsSet,
		}
```

Reply:
I didn't want the constraint with non-default clusters to work any differently than the default cluster (at least for this PR), and this it the constraint we set for the default cluster today. Arguably we want to keep the semver check so that we're ensuring we're only landing on compatible versions of Vault, but that hasn't been meaningful in a long time because we're using very stable interfaces.

That being said, it'd be worth checking that this version constraint is still valid for the Workload Identity login -- when did Vault get JWT login capability?

Reply:
> I didn't want the constraint with non-default clusters to work any differently than the default cluster (at least for this PR)

Makes sense 👍 

> when did Vault get JWT login capability?

0.10.4 it seems: https://github.com/hashicorp/vault/blob/main/CHANGELOG-v0.md#0104-july-25th-2018

Reply:
Ok, for now I'm going to merge this and then follow-up with a PR where we define an official minimal Vault version including documentation later on.